importerror due to cwd being part of sys.path

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->

### Steps to reproduce

```
> $ python3 -m venv pylint_venv                                                                                                                               
                                                                                                                                                               
> $ act pylint_venv                                                                                                                                           
activate virtualenv in pylint_venv
                                                                                                                                                               
(pylint_venv) > $ pip install toml pylint                                                                                                                     
Collecting toml
  Using cached https://files.pythonhosted.org/packages/a2/12/ced7105d2de62fa7c8fb5fce92cc4ce66b57c95fb875e9318dba7f8c5db0/toml-0.10.0-py2.py3-none-any.whl
Collecting pylint
  Using cached https://files.pythonhosted.org/packages/60/c2/b3f73f4ac008bef6e75bca4992f3963b3f85942e0277237721ef1c151f0d/pylint-2.3.1-py3-none-any.whl
Collecting mccabe<0.7,>=0.6 (from pylint)
  Using cached https://files.pythonhosted.org/packages/87/89/479dc97e18549e21354893e4ee4ef36db1d237534982482c3681ee6e7b57/mccabe-0.6.1-py2.py3-none-any.whl
Collecting isort<5,>=4.2.5 (from pylint)
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/1c/d9/bf5848b376e441ff358a14b954476423eeeb8c9b78c10074b7f53ce2918d/isort-4.3.20-py2.py3-none-any.whl
Collecting astroid<3,>=2.2.0 (from pylint)
  Using cached https://files.pythonhosted.org/packages/d5/ad/7221a62a2dbce5c3b8c57fd18e1052c7331adc19b3f27f1561aa6e620db2/astroid-2.2.5-py3-none-any.whl
Collecting wrapt (from astroid<3,>=2.2.0->pylint)
  Using cached https://files.pythonhosted.org/packages/23/84/323c2415280bc4fc880ac5050dddfb3c8062c2552b34c2e512eb4aa68f79/wrapt-1.11.2.tar.gz
Collecting six (from astroid<3,>=2.2.0->pylint)
  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl
Collecting typed-ast>=1.3.0; implementation_name == "cpython" (from astroid<3,>=2.2.0->pylint)
  Using cached https://files.pythonhosted.org/packages/fb/56/dd4e168a0009da85c78c6cfe91f5b2df2c7bbed60f3ba778c4a71289e6fb/typed_ast-1.4.0-cp37-cp37m-manylinux1_x86_64.whl
Collecting lazy-object-proxy (from astroid<3,>=2.2.0->pylint)
  Using cached https://files.pythonhosted.org/packages/43/a5/1b19b094ad19bce55b5b6d434020f5537b424fd2b3cff0fbef23d7bb5a95/lazy_object_proxy-1.4.1-cp37-cp37m-manylinux1_x86_64.whl
Building wheels for collected packages: wrapt
  Running setup.py bdist_wheel for wrapt ... error
  Complete output from command /tmp/pylint_venv/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-1jfcn_y2/wrapt/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-_5u2x0xx --python-tag cp37:
  usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: -c --help [cmd1 cmd2 ...]
     or: -c --help-commands
     or: -c cmd --help
  
  error: invalid command 'bdist_wheel'
  
  ----------------------------------------
  Failed building wheel for wrapt
  Running setup.py clean for wrapt
Failed to build wrapt
Installing collected packages: toml, mccabe, isort, wrapt, six, typed-ast, lazy-object-proxy, astroid, pylint
  Running setup.py install for wrapt ... done
Successfully installed astroid-2.2.5 isort-4.3.20 lazy-object-proxy-1.4.1 mccabe-0.6.1 pylint-2.3.1 six-1.12.0 toml-0.10.0 typed-ast-1.4.0 wrapt-1.11.2
                                                                                                                                                               
 
(pylint_venv) > $ cd ~/gitrepos/pytest/src/_pytest                                                                                                            
                                                                                                                                                                                                                                                                                                   
(pylint_venv) > $ pylint logging.py                                                                       
************* Module _pytest.logging
logging.py:85:2: W0511: TODO optimize this by introducing an option that tells the (fixme)
logging.py:265:0: C0301: Line too long (106/100) (line-too-long)
logging.py:303:0: C0301: Line too long (118/100) (line-too-long)
logging.py:305:0: C0301: Line too long (111/100) (line-too-long)
logging.py:308:0: C0301: Line too long (117/100) (line-too-long)
logging.py:320:0: C0301: Line too long (111/100) (line-too-long)
logging.py:324:0: C0301: Line too long (106/100) (line-too-long)
logging.py:327:0: C0301: Line too long (110/100) (line-too-long)
logging.py:338:0: C0301: Line too long (111/100) (line-too-long)
logging.py:342:0: C0301: Line too long (106/100) (line-too-long)
logging.py:451:0: C0330: Wrong hanging indentation before block (add 4 spaces).
            log_format
            ^   | (bad-continuation)
logging.py:538:0: C0330: Wrong hanging indentation before block (add 4 spaces).
            LogCaptureHandler(), formatter=self.formatter, level=self.log_level
            ^   | (bad-continuation)
logging.py:603:0: C0330: Wrong hanging indentation before block (add 4 spaces).
                        self.log_file_handler, level=self.log_file_level
                        ^   | (bad-continuation)
logging.py:646:0: C0301: Line too long (111/100) (line-too-long)
logging.py:649:0: C0301: Line too long (110/100) (line-too-long)
logging.py:6:0: E0401: Unable to import 'py' (import-error)
logging.py:22:28: E1101: Module 'logging' has no 'Formatter' member (no-member)
logging.py:28:8: E1101: Module 'logging' has no 'CRITICAL' member (no-member)
logging.py:29:8: E1101: Module 'logging' has no 'ERROR' member (no-member)
logging.py:30:8: E1101: Module 'logging' has no 'WARNING' member (no-member)
logging.py:31:8: E1101: Module 'logging' has no 'WARN' member (no-member)
logging.py:32:8: E1101: Module 'logging' has no 'INFO' member (no-member)
logging.py:33:8: E1101: Module 'logging' has no 'DEBUG' member (no-member)
logging.py:34:8: E1101: Module 'logging' has no 'NOTSET' member (no-member)
logging.py:40:29: W0212: Access to a protected member _fmt of a client class (protected-access)
logging.py:50:29: E1101: Module 'logging' has no 'getLevelName' member (no-member)
logging.py:62:4: C0111: Missing method docstring (missing-docstring)
logging.py:64:8: W0212: Access to a protected member _fmt of a client class (protected-access)
logging.py:22:0: R0903: Too few public methods (1/2) (too-few-public-methods)
logging.py:459:8: W0201: Attribute '_style' defined outside __init__ (attribute-defined-outside-init)
logging.py:68:28: E1101: Module 'logging' has no 'PercentStyle' member (no-member)
logging.py:81:4: C0111: Missing method docstring (missing-docstring)
logging.py:82:8: R1705: Unnecessary "else" after "return" (no-else-return)
logging.py:68:0: R0903: Too few public methods (1/2) (too-few-public-methods)
logging.py:95:0: C0111: Missing function docstring (missing-docstring)
logging.py:95:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
logging.py:108:51: W0622: Redefining built-in 'type' (redefined-builtin)
logging.py:191:18: E1101: Module 'logging' has no 'getLogger' member (no-member)
logging.py:216:24: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:221:8: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:227:8: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:229:4: C0111: Missing method docstring (missing-docstring)
logging.py:231:8: W0201: Attribute 'stream' defined outside __init__ (attribute-defined-outside-init)
logging.py:250:21: E1101: Module 'logging' has no 'getLogger' member (no-member)
logging.py:273:8: R1705: Unnecessary "else" after "return" (no-else-return)
logging.py:331:17: E1101: Module 'logging' has no 'getLogger' member (no-member)
logging.py:344:17: E1101: Module 'logging' has no 'getLogger' member (no-member)
logging.py:337:4: R0201: Method could be a function (no-self-use)
logging.py:366:4: W0212: Access to a protected member _finalize of a client class (protected-access)
logging.py:390:59: W0631: Using possibly undefined loop variable 'setting_name' (undefined-loop-variable)
logging.py:369:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
logging.py:396:0: C0111: Missing function docstring (missing-docstring)
logging.py:400:0: R0902: Too many instance attributes (11/7) (too-many-instance-attributes)
logging.py:424:34: E1101: Module 'logging' has no 'Formatter' member (no-member)
logging.py:430:36: E1101: Module 'logging' has no 'FileHandler' member (no-member)
logging.py:439:33: W0108: Lambda may not be necessary (unnecessary-lambda)
logging.py:457:24: E1101: Module 'logging' has no 'Formatter' member (no-member)
logging.py:459:8: W0212: Access to a protected member _style of a client class (protected-access)
logging.py:459:49: W0212: Access to a protected member _fmt of a client class (protected-access)
logging.py:459:49: W0212: Access to a protected member _style of a client class (protected-access)
logging.py:500:32: E1101: Module 'logging' has no 'FileHandler' member (no-member)
logging.py:514:4: C0111: Missing method docstring (missing-docstring)
logging.py:564:4: C0111: Missing method docstring (missing-docstring)
logging.py:569:4: C0111: Missing method docstring (missing-docstring)
logging.py:574:4: C0111: Missing method docstring (missing-docstring)
logging.py:579:4: C0111: Missing method docstring (missing-docstring)
logging.py:586:4: C0111: Missing method docstring (missing-docstring)
logging.py:591:4: C0111: Missing method docstring (missing-docstring)
logging.py:596:4: C0111: Missing method docstring (missing-docstring)
logging.py:614:4: C0111: Missing method docstring (missing-docstring)
logging.py:644:32: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:658:8: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:675:4: C0111: Missing method docstring (missing-docstring)
logging.py:692:12: E1101: Module 'logging' has no 'StreamHandler' member (no-member)
logging.py:684:16: W0201: Attribute '_first_record_emitted' defined outside __init__ (attribute-defined-outside-init)
logging.py:691:16: W0201: Attribute '_section_name_shown' defined outside __init__ (attribute-defined-outside-init)
Traceback (most recent call last):
  File "/tmp/pylint_venv/bin/pylint", line 10, in <module>
    sys.exit(run_pylint())
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/__init__.py", line 20, in run_pylint
    Run(sys.argv[1:])
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/lint.py", line 1628, in __init__
    linter.check(args)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/lint.py", line 943, in check
    self._do_check(files_or_modules)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/lint.py", line 1075, in _do_check
    self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/lint.py", line 1158, in check_astroid_module
    walker.walk(ast_node)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/utils.py", line 1305, in walk
    cb(astroid)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/checkers/imports.py", line 507, in leave_module
    std_imports, ext_imports, loc_imports = self._check_imports_order(node)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/pylint/checkers/imports.py", line 667, in _check_imports_order
    known_standard_library=self.config.known_standard_library,
  File "/tmp/pylint_venv/lib/python3.7/site-packages/isort/isort.py", line 57, in __init__
    self.config = settings.from_path(settings_path).copy()
  File "/tmp/pylint_venv/lib/python3.7/site-packages/isort/settings.py", line 177, in from_path
    _update_settings_with_config(path, 'pyproject.toml', [], ('tool.isort', ), computed_settings)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/isort/settings.py", line 207, in _update_settings_with_config
    _update_with_config_file(editor_config_file, sections, computed_settings)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/isort/settings.py", line 212, in _update_with_config_file
    settings = _get_config_data(file_path, sections).copy()
  File "/tmp/pylint_venv/lib/python3.7/site-packages/isort/settings.py", line 288, in _get_config_data
    config = toml.load(config_file)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/toml/decoder.py", line 110, in load
    if _ispath(f):
  File "/tmp/pylint_venv/lib/python3.7/site-packages/toml/decoder.py", line 29, in _ispath
    return _detect_pathlib_path(p)
  File "/tmp/pylint_venv/lib/python3.7/site-packages/toml/decoder.py", line 20, in _detect_pathlib_path
    import pathlib
  File "./pathlib.py", line 18, in <module>
    from pathlib import Path, PurePath
ImportError: cannot import name 'Path' from 'pathlib' (./pathlib.py)
```

### Current behavior

Pylint raises an exception if the current worrking directory contains a module names pathlib, that is different from the stdlib version of pathlib. This issue is probably not specific to the pathlib module, but applies to all modules from the stdlib that are shadowed by modules in the cwd that have the same name.
 
### Expected behavior

sys.path must not be modified to contain the CWD.

### pylint --version output

```
(pylint_venv) > $ pylint --version                                                                                                         
pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]
```

--- a/pylint/checkers/imports.py
+++ b/pylint/checkers/imports.py
@@ -776,7 +776,6 @@ class ImportsChecker(BaseChecker):
         except astroid.TooManyLevelsError:
             if _ignore_import_failure(importnode, modname, self._ignored_modules):
                 return None
-
             self.add_message("relative-beyond-top-level", node=importnode)
         except astroid.AstroidSyntaxError as exc:
             message = "Cannot import {!r} due to syntax error {!r}".format(

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -958,15 +958,22 @@ class PyLinter(
                 )
 
             filepath = files_or_modules[0]
-            self._check_files(
-                functools.partial(self.get_ast, data=_read_stdin()),
-                [self._get_file_descr_from_stdin(filepath)],
-            )
+            with fix_import_path(files_or_modules):
+                self._check_files(
+                    functools.partial(self.get_ast, data=_read_stdin()),
+                    [self._get_file_descr_from_stdin(filepath)],
+                )
         elif self.config.jobs == 1:
-            self._check_files(self.get_ast, self._iterate_file_descrs(files_or_modules))
+            with fix_import_path(files_or_modules):
+                self._check_files(
+                    self.get_ast, self._iterate_file_descrs(files_or_modules)
+                )
         else:
             check_parallel(
-                self, self.config.jobs, self._iterate_file_descrs(files_or_modules)
+                self,
+                self.config.jobs,
+                self._iterate_file_descrs(files_or_modules),
+                files_or_modules,
             )
 
     def check_single_file(self, name, filepath, modname):
@@ -1265,9 +1272,8 @@ class PyLinter(
         return note
 
 
-def check_parallel(linter, jobs, files):
-    """Use the given linter to lint the files with given amount of workers (jobs)
-    """
+def check_parallel(linter, jobs, files, arguments=None):
+    """Use the given linter to lint the files with given amount of workers (jobs)"""
     # The reporter does not need to be passed to worker processess, i.e. the reporter does
     # not need to be pickleable
     original_reporter = linter.reporter
@@ -1276,9 +1282,8 @@ def check_parallel(linter, jobs, files):
     # The linter is inherited by all the pool's workers, i.e. the linter
     # is identical to the linter object here. This is requred so that
     # a custom PyLinter object can be used.
-    with multiprocessing.Pool(
-        jobs, initializer=_worker_initialize, initargs=[linter]
-    ) as pool:
+    initializer = functools.partial(_worker_initialize, arguments=arguments)
+    with multiprocessing.Pool(jobs, initializer=initializer, initargs=[linter]) as pool:
         # ..and now when the workers have inherited the linter, the actual reporter
         # can be set back here on the parent process so that results get stored into
         # correct reporter
@@ -1311,7 +1316,7 @@ def check_parallel(linter, jobs, files):
 _worker_linter = None
 
 
-def _worker_initialize(linter):
+def _worker_initialize(linter, arguments=None):
     global _worker_linter  # pylint: disable=global-statement
     _worker_linter = linter
 
@@ -1319,6 +1324,9 @@ def _worker_initialize(linter):
     # parent process as _worker_check_file function's return value
     _worker_linter.set_reporter(reporters.CollectingReporter())
     _worker_linter.open()
+
+    # Patch sys.path so that each argument is importable just like in single job mode
+    _patch_sys_path(arguments or ())
 
 
 def _worker_check_single_file(file_item):
@@ -1450,6 +1458,21 @@ def preprocess_options(args, search_for):
             i += 1
 
 
+def _patch_sys_path(args):
+    original = list(sys.path)
+    changes = []
+    seen = set()
+    cwd = os.getcwd()
+    for arg in args:
+        path = utils.get_python_path(arg)
+        if path not in seen:
+            changes.append(path)
+            seen.add(path)
+
+    sys.path[:] = changes + sys.path
+    return original
+
+
 @contextlib.contextmanager
 def fix_import_path(args):
     """Prepare sys.path for running the linter checks.
@@ -1459,21 +1482,11 @@ def fix_import_path(args):
     We avoid adding duplicate directories to sys.path.
     `sys.path` is reset to its original value upon exiting this context.
     """
-    orig = list(sys.path)
-    changes = []
-    seen = set()
-    cwd = os.getcwd()
-    for arg in args:
-        path = utils.get_python_path(arg)
-        if path not in seen and path != cwd:
-            changes.append(path)
-            seen.add(path)
-
-    sys.path[:] = changes + sys.path
+    original = _patch_sys_path(args)
     try:
         yield
     finally:
-        sys.path[:] = orig
+        sys.path[:] = original
 
 
 class Run:
@@ -1754,11 +1767,8 @@ group are mutually exclusive.",
         # load plugin specific configuration.
         linter.load_plugin_configuration()
 
-        # insert current working directory to the python path to have a correct
-        # behaviour
-        with fix_import_path(args):
-            linter.check(args)
-            score_value = linter.generate_reports()
+        linter.check(args)
+        score_value = linter.generate_reports()
         if do_exit:
             if linter.config.exit_zero:
                 sys.exit(0)

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -1462,7 +1462,6 @@ def _patch_sys_path(args):
     original = list(sys.path)
     changes = []
     seen = set()
-    cwd = os.getcwd()
     for arg in args:
         path = utils.get_python_path(arg)
         if path not in seen:

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -1475,11 +1475,15 @@ def fix_import_path(args):
     """
     orig = list(sys.path)
     changes = []
+    seen = set()
+    cwd = os.getcwd()
     for arg in args:
         path = _get_python_path(arg)
-        if path not in changes:
+        if path not in seen and path != cwd:
             changes.append(path)
-    sys.path[:] = changes + ["."] + sys.path
+            seen.add(path)
+
+    sys.path[:] = changes + sys.path
     try:
         yield
     finally:

--- a/pylint/lint.py
+++ b/pylint/lint.py
@@ -114,20 +114,6 @@ def _get_new_args(message):
     return (message.msg_id, message.symbol, location, message.msg, message.confidence)
 
 
-def _get_python_path(filepath):
-    dirname = os.path.realpath(os.path.expanduser(filepath))
-    if not os.path.isdir(dirname):
-        dirname = os.path.dirname(dirname)
-    while True:
-        if not os.path.exists(os.path.join(dirname, "__init__.py")):
-            return dirname
-        old_dirname = dirname
-        dirname = os.path.dirname(dirname)
-        if old_dirname == dirname:
-            return os.getcwd()
-    return None
-
-
 def _merge_stats(stats):
     merged = {}
     by_msg = collections.Counter()
@@ -1478,7 +1464,7 @@ def fix_import_path(args):
     seen = set()
     cwd = os.getcwd()
     for arg in args:
-        path = _get_python_path(arg)
+        path = utils.get_python_path(arg)
         if path not in seen and path != cwd:
             changes.append(path)
             seen.add(path)

--- a/pylint/utils/utils.py
+++ b/pylint/utils/utils.py
@@ -4,12 +4,11 @@
 # For details: https://github.com/PyCQA/pylint/blob/master/COPYING
 
 import codecs
+import os
 import re
 import sys
 import textwrap
 import tokenize
-from os import linesep, listdir
-from os.path import basename, dirname, exists, isdir, join, normpath, splitext
 
 from astroid import Module, modutils
 
@@ -123,18 +122,18 @@ def expand_modules(files_or_modules, black_list, black_list_re):
     result = []
     errors = []
     for something in files_or_modules:
-        if basename(something) in black_list:
+        if os.path.basename(something) in black_list:
             continue
-        if _basename_in_blacklist_re(basename(something), black_list_re):
+        if _basename_in_blacklist_re(os.path.basename(something), black_list_re):
             continue
-        if exists(something):
+        if os.path.exists(something):
             # this is a file or a directory
             try:
                 modname = ".".join(modutils.modpath_from_file(something))
             except ImportError:
-                modname = splitext(basename(something))[0]
-            if isdir(something):
-                filepath = join(something, "__init__.py")
+                modname = os.path.splitext(os.path.basename(something))[0]
+            if os.path.isdir(something):
+                filepath = os.path.join(something, "__init__.py")
             else:
                 filepath = something
         else:
@@ -150,7 +149,7 @@ def expand_modules(files_or_modules, black_list, black_list_re):
                 errors.append({"key": "fatal", "mod": modname, "ex": ex})
                 continue
 
-        filepath = normpath(filepath)
+        filepath = os.path.normpath(filepath)
         modparts = (modname or something).split(".")
 
         try:
@@ -158,7 +157,7 @@ def expand_modules(files_or_modules, black_list, black_list_re):
         except ImportError:
             # Might not be acceptable, don't crash.
             is_namespace = False
-            is_directory = isdir(something)
+            is_directory = os.path.isdir(something)
         else:
             is_namespace = modutils.is_namespace(spec)
             is_directory = modutils.is_directory(spec)
@@ -176,16 +175,18 @@ def expand_modules(files_or_modules, black_list, black_list_re):
 
         has_init = (
             not (modname.endswith(".__init__") or modname == "__init__")
-            and basename(filepath) == "__init__.py"
+            and os.path.basename(filepath) == "__init__.py"
         )
 
         if has_init or is_namespace or is_directory:
             for subfilepath in modutils.get_module_files(
-                dirname(filepath), black_list, list_all=is_namespace
+                os.path.dirname(filepath), black_list, list_all=is_namespace
             ):
                 if filepath == subfilepath:
                     continue
-                if _basename_in_blacklist_re(basename(subfilepath), black_list_re):
+                if _basename_in_blacklist_re(
+                    os.path.basename(subfilepath), black_list_re
+                ):
                     continue
 
                 modpath = _modpath_from_file(subfilepath, is_namespace)
@@ -207,17 +208,19 @@ def register_plugins(linter, directory):
     'register' function in each one, used to register pylint checkers
     """
     imported = {}
-    for filename in listdir(directory):
-        base, extension = splitext(filename)
+    for filename in os.listdir(directory):
+        base, extension = os.path.splitext(filename)
         if base in imported or base == "__pycache__":
             continue
         if (
             extension in PY_EXTS
             and base != "__init__"
-            or (not extension and isdir(join(directory, base)))
+            or (not extension and os.path.isdir(os.path.join(directory, base)))
         ):
             try:
-                module = modutils.load_module_from_file(join(directory, filename))
+                module = modutils.load_module_from_file(
+                    os.path.join(directory, filename)
+                )
             except ValueError:
                 # empty module name (usually emacs auto-save files)
                 continue
@@ -323,7 +326,7 @@ def _check_csv(value):
 def _comment(string):
     """return string as a comment"""
     lines = [line.strip() for line in string.splitlines()]
-    return "# " + ("%s# " % linesep).join(lines)
+    return "# " + ("%s# " % os.linesep).join(lines)
 
 
 def _format_option_value(optdict, value):

--- a/pylint/utils/utils.py
+++ b/pylint/utils/utils.py
@@ -115,6 +115,20 @@ def _modpath_from_file(filename, is_namespace):
     )
 
 
+def get_python_path(filepath):
+    dirname = os.path.realpath(os.path.expanduser(filepath))
+    if not os.path.isdir(dirname):
+        dirname = os.path.dirname(dirname)
+    while True:
+        if not os.path.exists(os.path.join(dirname, "__init__.py")):
+            return dirname
+        old_dirname = dirname
+        dirname = os.path.dirname(dirname)
+        if old_dirname == dirname:
+            return os.getcwd()
+    return None
+
+
 def expand_modules(files_or_modules, black_list, black_list_re):
     """take a list of files/modules/packages and return the list of tuple
     (file, module name) which have to be actually checked

--- a/pylint/utils/utils.py
+++ b/pylint/utils/utils.py
@@ -135,15 +135,22 @@ def expand_modules(files_or_modules, black_list, black_list_re):
     """
     result = []
     errors = []
+    path = sys.path.copy()
+
     for something in files_or_modules:
         if os.path.basename(something) in black_list:
             continue
         if _basename_in_blacklist_re(os.path.basename(something), black_list_re):
             continue
+
+        module_path = get_python_path(something)
+        additional_search_path = [module_path] + path
         if os.path.exists(something):
             # this is a file or a directory
             try:
-                modname = ".".join(modutils.modpath_from_file(something))
+                modname = ".".join(
+                    modutils.modpath_from_file(something, path=additional_search_path)
+                )
             except ImportError:
                 modname = os.path.splitext(os.path.basename(something))[0]
             if os.path.isdir(something):
@@ -154,7 +161,9 @@ def expand_modules(files_or_modules, black_list, black_list_re):
             # suppose it's a module or package
             modname = something
             try:
-                filepath = modutils.file_from_modpath(modname.split("."))
+                filepath = modutils.file_from_modpath(
+                    modname.split("."), path=additional_search_path
+                )
                 if filepath is None:
                     continue
             except (ImportError, SyntaxError) as ex:
@@ -167,7 +176,9 @@ def expand_modules(files_or_modules, black_list, black_list_re):
         modparts = (modname or something).split(".")
 
         try:
-            spec = modutils.file_info_from_modpath(modparts, path=sys.path)
+            spec = modutils.file_info_from_modpath(
+                modparts, path=additional_search_path
+            )
         except ImportError:
             # Might not be acceptable, don't crash.
             is_namespace = False

--- a/tests/test_self.py
+++ b/tests/test_self.py
@@ -664,3 +664,35 @@ class TestRunTC(object):
             ],
             code=0,
         )
+
+    def test_do_not_import_files_from_local_directory(self, tmpdir):
+        p_astroid = tmpdir / "astroid.py"
+        p_astroid.write("'Docstring'\nimport completely_unknown\n")
+        p_hmac = tmpdir / "hmac.py"
+        p_hmac.write("'Docstring'\nimport completely_unknown\n")
+
+        with tmpdir.as_cwd():
+            subprocess.check_output(
+                [
+                    sys.executable,
+                    "-m",
+                    "pylint",
+                    "astroid.py",
+                    "--disable=import-error,unused-import",
+                ],
+                cwd=str(tmpdir),
+            )
+
+        # Test with multiple jobs
+        with tmpdir.as_cwd():
+            subprocess.call(
+                [
+                    sys.executable,
+                    "-m",
+                    "pylint",
+                    "-j2",
+                    "hmac.py",
+                    "--disable=import-error,unused-import",
+                ],
+                cwd=str(tmpdir),
+            )

--- a/tests/test_self.py
+++ b/tests/test_self.py
@@ -683,7 +683,22 @@ class TestRunTC(object):
                 cwd=str(tmpdir),
             )
 
-        # Test with multiple jobs
+        # Linting this astroid file does not import it
+        with tmpdir.as_cwd():
+            subprocess.check_output(
+                [
+                    sys.executable,
+                    "-m",
+                    "pylint",
+                    "-j2",
+                    "astroid.py",
+                    "--disable=import-error,unused-import",
+                ],
+                cwd=str(tmpdir),
+            )
+
+        # Test with multiple jobs for hmac.py for which we have a
+        # CVE against: https://github.com/PyCQA/pylint/issues/959
         with tmpdir.as_cwd():
             subprocess.call(
                 [

--- a/tests/unittest_lint.py
+++ b/tests/unittest_lint.py
@@ -165,7 +165,7 @@ def fake_path():
 
 def test_no_args(fake_path):
     with lint.fix_import_path([]):
-        assert sys.path == ["."] + fake_path
+        assert sys.path == fake_path
     assert sys.path == fake_path
 
 
@@ -175,7 +175,7 @@ def test_no_args(fake_path):
 def test_one_arg(fake_path, case):
     with tempdir() as chroot:
         create_files(["a/b/__init__.py"])
-        expected = [join(chroot, "a")] + ["."] + fake_path
+        expected = [join(chroot, "a")] + fake_path
 
         assert sys.path == fake_path
         with lint.fix_import_path(case):
@@ -195,7 +195,7 @@ def test_one_arg(fake_path, case):
 def test_two_similar_args(fake_path, case):
     with tempdir() as chroot:
         create_files(["a/b/__init__.py", "a/c/__init__.py"])
-        expected = [join(chroot, "a")] + ["."] + fake_path
+        expected = [join(chroot, "a")] + fake_path
 
         assert sys.path == fake_path
         with lint.fix_import_path(case):
@@ -214,14 +214,10 @@ def test_two_similar_args(fake_path, case):
 def test_more_args(fake_path, case):
     with tempdir() as chroot:
         create_files(["a/b/c/__init__.py", "a/d/__init__.py", "a/e/f.py"])
-        expected = (
-            [
-                join(chroot, suffix)
-                for suffix in [sep.join(("a", "b")), "a", sep.join(("a", "e"))]
-            ]
-            + ["."]
-            + fake_path
-        )
+        expected = [
+            join(chroot, suffix)
+            for suffix in [sep.join(("a", "b")), "a", sep.join(("a", "e"))]
+        ] + fake_path
 
         assert sys.path == fake_path
         with lint.fix_import_path(case):

